File: packages/coordinator/src/agui.ts
/**
 * P11.5.11 (H25) — AG-UI wire protocol (doc 50): tool calls + UI updates over
 * ONE JSON channel (~16 event types). This is the typed envelope + codec that
 * sits on top of the existing P0.5 framed IPC — the coordinator encodes
 * events, the UI decodes them, and generative UI payloads ride inside
 * `artifact` events.
 *
 * Live transport (the framed-IPC hop):
 * - coordinator → Rust → UI: [`notifyAgui`] emits an `agui/event`
 *   notification (`params.line` = the encoded envelope); the Rust relay
 *   forwards it to the Tauri `agui-event` emit.
 * - UI → coordinator: the Tauri `agui_send` command writes an `agui/event`
 *   notification back; [`dispatchAguiLine`] routes it to the handlers
 *   registered via [`onAguiEvent`] (e.g. `interrupt_resolved`).
 */

import { notify } from "./frame";

/** The AG-UI event set (stable names, snake_case over the wire). */
export type AguiEventType =
  | "user_message_created"
  | "assistant_message_created"
  | "assistant_message_delta"
  | "tool_call_created"
  | "tool_call_delta"
  | "tool_call_result"
  | "agent_message"
  | "agent_state_changed"
  | "artifact_created"
  | "artifact_updated"
  | "interrupt_created"
  | "interrupt_resolved"
  | "session_created"
  | "session_updated"
  | "error"
  | "done";

export interface AguiEnvelope<T = unknown> {
  type: AguiEventType;
  /** Correlation id (maps to the framed IPC message id). */
  id: string;
  /** ISO timestamp. */
  ts: string;
  data: T;
}

export interface ToolCallData {
  call_id: string;
  name: string;
  args: Record<string, unknown>;
  /** session/progress state for the tool (UI renders this live). */
  state?: "pending" | "running" | "done" | "error";
}

export interface ArtifactData {
  artifact_id: string;
  /** Version bumps on every artifact_updated (make-live version selector). */
  version: number;
  kind: "html" | "mermaid" | "descriptor" | "markdown";
  payload: unknown;
}

export interface InterruptData {
  interrupt_id: string;
  kind: "permission" | "mcq" | "diff" | "budget";
  title: string;
  description: string;
  options: { label: string; value: string }[];
}

const TYPES: readonly AguiEventType[] = [
  "user_message_created",
  "assistant_message_created",
  "assistant_message_delta",
  "tool_call_created",
  "tool_call_delta",
  "tool_call_result",
  "agent_message",
  "agent_state_changed",
  "artifact_created",
  "artifact_updated",
  "interrupt_created",
  "interrupt_resolved",
  "session_created",
  "session_updated",
  "error",
  "done",
];

/** Encode an event for the wire (JSON, one line — NDJSON over the frame). */
export function encodeAgui<T>(type: AguiEventType, id: string, data: T): string {
  const env: AguiEnvelope<T> = { type, id, ts: new Date().toISOString(), data };
  return JSON.stringify(env);
}

/** Decode one wire line; returns null on malformed/unknown type (never throws). */
export function decodeAgui(line: string): AguiEnvelope | null {
  try {
    const parsed = JSON.parse(line) as AguiEnvelope;
    if (typeof parsed.type !== "string" || !TYPES.includes(parsed.type as AguiEventType)) {
      return null;
    }
    if (typeof parsed.id !== "string" || typeof parsed.ts !== "string" || !("data" in parsed)) {
      return null;
    }
    return parsed;
  } catch {
    return null;
  }
}

/** Emit one AG-UI event over the framed IPC as an `agui/event` notification. */
export function notifyAgui<T>(type: AguiEventType, id: string, data: T): void {
  notify("agui/event", { line: encodeAgui(type, id, data) });
}

type AguiHandler = (env: AguiEnvelope) => void;

const handlers = new Map<AguiEventType, AguiHandler[]>();

/** Register a handler for inbound UI→coordinator AG-UI events (e.g. answers
 * to `interrupt_created` arrive as `interrupt_resolved`). */
export function onAguiEvent(type: AguiEventType, handler: AguiHandler): () => void {
  const list = handlers.get(type) ?? [];
  list.push(handler);
  handlers.set(type, list);
  return () => {
    const cur = handlers.get(type) ?? [];
    handlers.set(
      type,
      cur.filter((h) => h !== handler),
    );
  };
}

/** Route one inbound wire line to its handlers. Returns true when at least
 * one handler ran (false = unknown type / no listener / malformed). */
export function dispatchAguiLine(line: string): boolean {
  const env = decodeAgui(line);
  if (!env) return false;
  const list = handlers.get(env.type);
  if (!list || list.length === 0) return false;
  for (const h of list) h(env);
  return true;
}

/** Narrowing helpers. */
export function isToolCall(e: AguiEnvelope | null): e is AguiEnvelope<ToolCallData> {
  return e !== null && (e.type === "tool_call_created" || e.type === "tool_call_result");
}

export function isArtifact(e: AguiEnvelope | null): e is AguiEnvelope<ArtifactData> {
  return e !== null && (e.type === "artifact_created" || e.type === "artifact_updated");
}

export function isInterrupt(e: AguiEnvelope | null): e is AguiEnvelope<InterruptData> {
  return e !== null && (e.type === "interrupt_created" || e.type === "interrupt_resolved");
}

File: crates/everyaios-core/src/agui.rs
//! AG-UI live transport (P11.5.11 — H25, doc 50). The Rust twin of the
//! coordinator's `agui.ts` codec: tool calls + UI updates ride ONE JSON
//! channel (~16 event types), with generative-UI payloads inside `artifact`
//! events.
//!
//! Wire contract (matches `agui.ts` byte-for-byte):
//! ```json
//! { "type": "tool_call_created", "id": "c1", "ts": "<iso>", "data": {…} }
//! ```
//! - coordinator → Rust → UI: the coordinator notifies `agui/event`
//!   (`params.line` = the encoded envelope line); the relay forwards it to
//!   the `on_agui` sink → Tauri emits `agui-event` with the raw line.
//! - UI → coordinator: the Tauri `agui_send` command writes an `agui/event`
//!   notification to the sidecar; the coordinator dispatches it (e.g.
//!   `interrupt_resolved` answers an outstanding AG-UI interrupt).

use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

/// The AG-UI envelope (wire shape identical to the TS `AguiEnvelope`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct AguiEnvelope {
    /// One of the 16 AG-UI event types (snake_case over the wire).
    pub r#type: String,
    /// Correlation id (maps to the framed IPC message id).
    pub id: String,
    /// ISO timestamp.
    pub ts: String,
    /// Event payload (tool calls, artifacts, interrupts, …).
    pub data: serde_json::Value,
}

/// Encode one envelope into a single JSON line (the `agui/event` wire unit).
pub fn encode_agui(r#type: &str, id: &str, ts: &str, data: serde_json::Value) -> String {
    serde_json::to_string(&AguiEnvelope {
        r#type: r#type.to_string(),
        id: id.to_string(),
        ts: ts.to_string(),
        data,
    })
    .unwrap_or_else(|_| r#"{"type":"error","id":"","ts":"","data":{}}"#.to_string())
}

/// Decode one `agui/event` line; `None` on malformed input (forward-compat:
/// an unknown event type still decodes — the `type` is an open string).
pub fn decode_agui(line: &str) -> Option<AguiEnvelope> {
    serde_json::from_str(line).ok()
}

/// Convenience envelope constructors (test + relay ergonomics).
pub fn envelope(r#type: &str, id: &str, ts: &str, data: serde_json::Value) -> AguiEnvelope {
    AguiEnvelope {
        r#type: r#type.into(),
        id: id.into(),
        ts: ts.into(),
        data,
    }
}

/// The UI event sink: receives raw encoded envelope lines.
type AguiSink = Box<dyn Fn(String) + Send>;

/// Forwards `agui/event` lines to the UI (Tauri emits `agui-event`).
#[derive(Clone, Default)]
pub struct AguiRelay {
    on_event: Arc<Mutex<Option<AguiSink>>>,
}

impl AguiRelay {
    pub fn new() -> Self {
        Self::default()
    }

    /// Attach (or replace) the UI sink. Call once at boot from the Tauri shell.
    pub fn attach(&self, sink: impl Fn(String) + Send + 'static) {
        *self.on_event.lock().unwrap_or_else(|e| e.into_inner()) = Some(Box::new(sink));
    }

    /// Forward one encoded envelope line to the UI (fire-and-forget).
    pub fn forward(&self, line: &str) {
        if let Some(sink) = self.on_event.lock().unwrap_or_else(|e| e.into_inner()).as_ref() {
            sink(line.to_string());
        }
    }

    /// Forward one envelope (encode-then-forward).
    pub fn forward_envelope(&self, env: &AguiEnvelope) {
        self.forward(&serde_json::to_string(env).unwrap_or_default());
    }

    /// True when a sink is attached (the UI is listening).
    pub fn is_attached(&self) -> bool {
        self.on_event
            .lock()
            .unwrap_or_else(|e| e.into_inner())
            .is_some()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn envelope_roundtrip_matches_wire_shape() {
        let env = envelope(
            "tool_call_created",
            "c1",
            "2026-08-23T00:00:00Z",
            serde_json::json!({ "callId": "t1", "name": "read_file", "args": { "path": "a.rs" } }),
        );
        let line = serde_json::to_string(&env).unwrap();
        let decoded = decode_agui(&line).expect("round-trips");
        assert_eq!(decoded, env);
        assert_eq!(decoded.r#type, "tool_call_created");
        assert_eq!(decoded.data["name"], "read_file");
    }

    #[test]
    fn encode_agui_produces_decodable_line() {
        let line = encode_agui("artifact_created", "c2", "t", serde_json::json!({ "v": 1 }));
        let env = decode_agui(&line).expect("decodes");
        assert_eq!(env.r#type, "artifact_created");
        assert_eq!(env.id, "c2");
    }

    #[test]
    fn malformed_line_decodes_to_none() {
        assert!(decode_agui("not json").is_none());
        assert!(decode_agui("").is_none());
        // Missing `data` → not a valid envelope (matches the TS codec, where
        // `data` is required). Unknown *types* still decode (open string).
        assert!(decode_agui(r#"{"type":"done","id":"x","ts":"t"}"#).is_none());
    }

    #[test]
    fn relay_forwards_only_when_attached() {
        let relay = AguiRelay::new();
        let got = Arc::new(Mutex::new(Vec::new()));
        let got2 = Arc::clone(&got);
        relay.attach(move |line| got2.lock().unwrap().push(line));
        relay.forward("{\"type\":\"done\"}");
        relay.forward_envelope(&envelope("session_created", "s1", "t", serde_json::json!({})));
        let lines = got.lock().unwrap();
        assert_eq!(lines.len(), 2);
        assert!(lines[1].contains("session_created"));
    }

    #[test]
    fn unknown_event_types_still_decode() {
        let line = encode_agui("brand_new_event_kind", "x", "t", serde_json::json!({ "a": 1 }));
        let env = decode_agui(&line).unwrap();
        assert_eq!(env.r#type, "brand_new_event_kind");
    }
}
